p2p/libp2p/defineOptions.go

```diff
@@ -0,0 +1,49 @@
+package libp2p
+
+import (
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/p2p"
+)
+
+var readTimeout = time.Second * 10
+
+// Option represents a functional configuration parameter that can operate
+//  over the networkMessenger struct.
+type Option func(*networkMessenger) error
+
+// WithAuthentication sets up the authentication mechanism and peer id - public key collection
+func WithAuthentication(
+	networkShardingCollector p2p.NetworkShardingCollector,
+	signerVerifier p2p.SignerVerifier,
+	marshalizer p2p.Marshalizer,
+) Option {
+	return func(mes *networkMessenger) error {
+		if check.IfNil(networkShardingCollector) {
+			return p2p.ErrNilNetworkShardingCollector
+		}
+		if check.IfNil(signerVerifier) {
+			return p2p.ErrNilSignerVerifier
+		}
+		if check.IfNil(marshalizer) {
+			return p2p.ErrNilMarshalizer
+		}
```

Comment:
do the tests in WithAuthentication not in the returned function

Reply:
+1

Reply:
Can not be done, as the "with options" pattern should return only an Option not an (Option, error) tuple. So the checking can not be moved outside the anonymous function.